### ISSUE TITLE
[Training] How to update running_mean and running_var of BatchNormalization during training

### ISSUE THREAD

**Issue by hbwx24** [microsoft/onnxruntime#19370]
### Describe the issue

I want to train a model with BatchNormalization. But it was found that the mean and var of BN could not be updated. How to update mean and var of BN?

- In addition, I have a very important question related to my code development.
  I found that the API of Module requires eval_model and train_model. During training, does the forward process use the eval_model, and the backward process uses the train_model? If so, I found that the `training_mode` of the BN of the eval model must be 0, does that mean not Support updating mean and var?

`class onnxruntime.training.api.Module(train_model_uri: [PathLike](https://docs.python.org/3/library/os.html#os.PathLike), state: [CheckpointState](https://onnxruntime.ai/docs/api/python/on_device_training/training_api.html#onnxruntime.training.api.CheckpointState), eval_model_uri: [Optional](https://docs.python.org/3/library/typing.html#typing.Optional)[[PathLike](https://docs.python.org/3/library/os.html#os.PathLike)] = None, device: [str](https://docs.python.org/3/library/stdtypes.html#str) = 'cpu', session_options: [Optional](https://docs.python.org/3/library/typing.html#typing.Optional)[[SessionOptions](https://onnxruntime.ai/docs/api/python/api_summary.html#onnxruntime.SessionOptions)] = None)[[source]](https://onnxruntime.ai/docs/api/python/modules/onnxruntime/training/api/module.html#Module)`
Module

### To reproduce

![eval-model](https://github.com/microsoft/onnxruntime/assets/22128369/89459cad-ac5e-4746-996a-3a5d568981a4)
![training-model](https://github.com/microsoft/onnxruntime/assets/22128369/477faf77-6fa2-41a6-9c1e-6fad589a533b)

### ONNX Runtime Installation

Built from Source

### ONNX Runtime Version or Commit ID

 1.16.3

### PyTorch Version

1.10

### Execution Provider

CUDA

**Comment by hbwx24** [comment]
If the current API needs to be improved to meet the above needs, I will be very happy if you can tell me about the design ideas of the ORT code. Because it will be more convenient for me to edit the source code.

**Comment by hbwx24** [comment]
If the current API needs to be improved to meet the above needs, I will be very happy if you can tell me about the design ideas of the ORT code. Because it will be more convenient for me to edit the source code.

**Comment by baijumeswani** [comment]
Sorry for the late response. 

> In addition, I have a very important question related to my code development.
I found that the API of Module requires eval_model and train_model. During training, does the forward process use the eval_model, and the backward process uses the train_model? If so, I found that the training_mode of the BN of the eval model must be 0, does that mean not Support updating mean and var?

Training model contains the forward+loss+backward graph where as the eval model contains only the forward+loss graph. The training model is strictly used during model training (forward+loss+gradient computation) where as the eval model is strictly used during model eval (forward+loss computation).

During training, the running mean and variance will be updated for BN with the training model but not with the eval model. This is precisely what `training_mode` flag is meant for.

**Comment by hbwx24** [comment]
> Sorry for the late response.
> 
> > In addition, I have a very important question related to my code development.
> > I found that the API of Module requires eval_model and train_model. During training, does the forward process use the eval_model, and the backward process uses the train_model? If so, I found that the training_mode of the BN of the eval model must be 0, does that mean not Support updating mean and var?
> 
> Training model contains the forward+loss+backward graph where as the eval model contains only the forward+loss graph. The training model is strictly used during model training (forward+loss+gradient computation) where as the eval model is strictly used during model eval (forward+loss computation).
> 
> During training, the running mean and variance will be updated for BN with the training model but not with the eval model. This is precisely what `training_mode` flag is meant for.

Thanks for your

**Comment by hbwx24** [comment]
Thank you for your reply.

> Sorry for the late response.
> 
> Training model contains the forward+loss+backward graph where as the eval model contains only the forward+loss graph. The training model is strictly used during model training (forward+loss+gradient computation) where as the eval model is strictly used during model eval (forward+loss computation).
> 
> During training, the running mean and variance will be updated for BN with the training model but not with the eval model. This is precisely what `training_mode` flag is meant for.

I have two questions：

#### 1. As answered above, does the following code that backwards and sets the gradient only use the training model?
```python
module = Module(train_model_new,
                    state,
                    eval_model_new,
                    device=device,
                    session_options=session_options)
for _ in range(10):
    module.train()
    training_loss = module(data)
    optimizer.step()
    module.lazy_reset_grad()

module.eval()
module.export_model_for_inferencing(final_output,graph_output_names=graph_output_names)
```

#### 2. Both eval model and traning model generated by `artifacts.generate_artifacts`. After executing backward, the inference model is exported by `export_model_for_inferencing`. At this time, it is found that the scale/bias of BN are updated, but the mean and var are not updated. Why is mean/var not updated? Is it because my code has missed some settings?

- origin model
  
![image](https://github.com/microsoft/onnxruntime/assets/22128369/a378ae6a-321f-4787-92d1-b4e00a62378e)

![image](https://github.com/microsoft/onnxruntime/assets/22128369/84d9722c-3cf4-4d30-a6df-1ffac0af82cb)

- eval model
  
- training model：BN output runing mean/var, but not update running mean/var in onnx model.
  
![image](https://github.com/microsoft/onnxruntime/assets/22128369/4371b9e4-7701-4b85-8092-328aef6bf0d6)


- latest eval model generated by `export_model_for_inferencing`.
  
![image](https://github.com/microsoft/onnxruntime/assets/22128369/c4317dfd-5357-4c55-bf1e-0f3a6668c85e)

**Comment by baijumeswani** [comment]
> 1. As answered above, does the following code that backwards and sets the gradient only use the training model?

Yes, you are right. The backward computation only uses the training model.

> 2. Both eval model and traning model generated by artifacts.generate_artifacts. After executing backward, the inference model is exported by export_model_for_inferencing. At this time, it is found that the scale/bias of BN are updated, but the mean and var are not updated. Why is mean/var not updated? Is it because my code has missed some settings?

Are you able to share the script you're using to generate the training artifacts? Particularly what you're setting as requires_grad and frozen_params?
